### PR TITLE
persist: add new persistent source example

### PR DESCRIPTION
**Pull request by aljoscha**
Not necessarily meant for merging, more as a discussion starter. This is still rough around the edges.

### CODE REVIEW

File: src/persist-client/examples/persistent_source.rs
Comment (danhhz) [inline src/persist-client/examples/persistent_source.rs:220]: why the -1? upper is exclusive, right? also given that lower is set to self.count the next time through, don't the lowers+uppers of the two batches not line up?

File: src/persist-client/examples/persistent_source.rs
Comment (danhhz) [inline src/persist-client/examples/persistent_source.rs:470]: nit: why does read_batch return an Option? this loop is simpler if it always waits until it can return a batch

File: src/persist-client/examples/persistent_source.rs
Comment (danhhz) [inline src/persist-client/examples/persistent_source.rs:422]: do we think the real system will be routing all the source data through the timestamper or would the real source have an API more like 1) notify me of new data which announces some sort of token/index (for the timestamper) 2) fetch new data based the token/index having been routed through timestamping?

it's 100% fine if this example is simplified for illustrative purposes, but if so, let's leave a comment to that effect

File: src/persist-client/examples/persistent_source.rs
Comment (danhhz) [inline src/persist-client/examples/persistent_source.rs:527]: can we use the language from the reclocking design doc for the various applicable pieces? e.g. remap, reclocked, reclocking?

File: src/persist-client/examples/persistent_source.rs
Comment (danhhz) [inline src/persist-client/examples/persistent_source.rs:38]: I came to regret my suggestion to structure this as a "cargo example" last time since that means it's a separate binary to be linked. if it's not too much trouble, how about making this a `#[test]` of some sort even if it doesn't really have assertions. and then since the runtime will be long from the sleeps, let's toss on an `#[ignore]` too so it has to be specifically requested

**Review by danhhz — COMMENTED** [review]
this is awesome! it's already doing its job given that I have a few questions below about what our actual API for sources will be and about concretely what data should be flowing along various channels

since the major stakeholders here will be the storage team, perhaps it'd be better for them to be the primary reviewers instead of ruchir and I? @elindsey do we have any sense yet of who on storage would be the best person to start pinging about stuff like this?

File: src/persist-client/examples/persistent_source.rs
Comment (aljoscha) [reply src/persist-client/examples/persistent_source.rs:220]: It's because I just wanted to get this to work without thinking too hard. The problem is that for determining timestamp bindings, we need an inclusive "upper", that is, just the highest offset in that batch (for Kafka).

And yes, this doesn't mesh well with how differential/timely use lower/upper. We might have to use separate terms/concepts here.

File: src/persist-client/examples/persistent_source.rs
Comment (aljoscha) [reply src/persist-client/examples/persistent_source.rs:470]: I had a thing in my head where sources would say "I don't have data right now, but I might in the future", and the `Option` is communicating that. Though in an async/async task world that might not be necessary.

File: src/persist-client/examples/persistent_source.rs
Comment (aljoscha) [reply src/persist-client/examples/persistent_source.rs:422]: It's more for illustrative purposes but I don't yet have an idea on how the real thing would work. The source could just, when you read a batch, also return the offset uppers, and we notify the timestamping logic so that it can start minting early.

File: src/persist-client/examples/persistent_source.rs
Comment (aljoscha) [reply src/persist-client/examples/persistent_source.rs:527]: 👌

File: src/persist-client/examples/persistent_source.rs
Comment (aljoscha) [reply src/persist-client/examples/persistent_source.rs:38]: can do!

File: src/persist-client/examples/persistent_source.rs
Comment (aljoscha) [reply src/persist-client/examples/persistent_source.rs:220]: Concretely, when a batch contains updates up to offset `t`, the upper (in differential world) would be `t+1`. But I'm only interested in minting a binding up to `t` because there probably is no data yet under `t+1`.

Also keep in mind that + and - might not always be easy to do on these "offset" types, _maybe_. Also, the examples use offsets, while Kafka would have `(partition, offset)`, but the ideas map 1-to-1.

**Comment by aljoscha** [comment]
@danhhz updated with a "real" `IteratedConsensus` interface that should be quite close to what we were thinking of exposing from persist.

((also addressed the other comments))

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:205]: sorry to nit but please let's not call this `append`, which is already a thing :). I'm fine with something like `compare_and_set`

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:192]: checking my understanding: a real impl of this would be backed by a persist shard (or rather a storage collection), right? as opposed to directly using persist's `Consensus` trait

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:551]: I didn't even know this syntax!

is it tricky to break out this loop into a method so these are just returns?

**Review by danhhz — COMMENTED** [review]
cool!

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:193]: Yes! I'm adding a comment that mentions this.

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:552]: Yeah, I'm not happy with this method, but didn't yet take any time to make it pretty.

**Comment by aljoscha** [comment]
@danhhz @ruchirK  This is now updated to "the real thing". There's a source pipeline that writes into a `WriteHandle`. There's a consumer pipeline that reads using a `ReadHandle`. And it uses the newly added `compare_and_append` to mint timestamp bindings with consensus (backed by `compare_and_append` on `WriteHandle`).

There's another thing that this surfaced again: we don't allow multiple writers to independently write their contiguous batches with regular `append`/`write_batch`. Instead we always check the global upper. I know we discussed this before but not sure where we'll land. It would definitely be easier for the source if every writer could just write without worrying about the other writers.

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:104]: confirmed offline that this is a bug in `Shard::write_batch`! feel free to fix it for me or wait for #11460 to land

**Review by danhhz — APPROVED** [review]
this is awesome! left a new more comments on the latest version, but it's definitely time to start getting some storage folks on here for input

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:24]: yes :sweat_smile:

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:149]: Nope, removing this.

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:232]: Yeah, I have to straighten this part out. I'm not sure it works correctly for timestamps that don't have a total order. That's partially why the example uses `u64`. :sweat_smile:

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:375]: Yep, updating this. The state is now shared in a process by sharing a Timestamper in an `Arc`, makes it more explicit what's happening from the outside.

**Review by ruchirK — COMMENTED** [review]
still reading through this but wanted to leave my one comment before i forgot

**Comment by ruchirK** [comment]
I tweaked this example locally to support multiple readers and found a strange behavior - it works fine with 40 readers, but gets completely stuck with 42 readers. This is surprising (to me at least) given that everything is running locally and there aren't a lot of writes so going to continue to debug this.

**Comment by ruchirK** [comment]
diff to make this multireader

```
diff --git a/src/persist-client/src/examples/persistent_source_example.rs b/src/persist-client/src/examples/persistent_source_example.rs
index 473aa55b1..dcdb0a8e9 100644
--- a/src/persist-client/src/examples/persistent_source_example.rs
+++ b/src/persist-client/src/examples/persistent_source_example.rs
@@ -37,6 +37,7 @@ use crate::Client;
 use crate::Id;
 use crate::Location;
 use crate::NO_TIMEOUT;
+use crate::ReadHandle;

 use self::impls::ConsensusTimestamper;
 use self::impls::PersistConsensus;
@@ -71,7 +72,8 @@ async fn run() -> Result<(), Box<dyn Error>> {
     let timestamper =
         ConsensusTimestamper::new(now_fn.clone(), timestamp_interval.clone(), consensus);

-    let (data_write, data_read) = persist.open(NO_TIMEOUT, data_id).await?;
+    let (data_write, _) = persist.open(NO_TIMEOUT, data_id).await?;
+    let num_readers = 42;

     // First, render one instance of the source.
     let source1 = source.clone();
@@ -89,18 +91,26 @@ async fn run() -> Result<(), Box<dyn Error>> {
         }
     });

-    let reader_pipeline1 = mz_ore::task::spawn(|| "reader-1", async move {
-        if let Err(e) = reader::spawn_reader_pipeline(
-            "reader-1".to_string(),
-            data_read,
-            Antichain::from_elem(Timestamp::new(0)),
-        )
-        .await
-        {
-            error!("error in reader: {:?}", e);
-        }
-    });
+    let mut reader_pipelines = vec![];
+    for i in 0..num_readers {
+        let reader_name = format!("reader-{}", i);
+        let reader_name_clone = reader_name.clone();
+        let data_read: ReadHandle<String, (), _, _> = persist
+            .open(NO_TIMEOUT, data_id).await?.1;
+        let pipeline = mz_ore::task::spawn(|| &reader_name_clone, async move {
+            if let Err(e) = reader::spawn_reader_pipeline(
+                reader_name,
+                data_read,
+                Antichain::from_elem(Timestamp::new(0)),
+            )
+            .await
+            {
+                error!("error in reader: {:?}", e);
+            }
+        });

+        reader_pipelines.push(pipeline);
+    }
     // TODO(aljoscha): We don't currently support multiple writers writing the same (definite) data
     // concurrently.

@@ -133,7 +143,9 @@ async fn run() -> Result<(), Box<dyn Error>> {
     source_pipeline1.await?;
     // source_pipeline2.await?;

-    reader_pipeline1.await?;
+    for pipeline in reader_pipelines.into_iter() {
+        pipeline.await?;
+    }

     Ok(())
 }
 ```

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:230]: It is an invariant hat has to be true, but this struct currently doesn't enforce/assert on it.

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:314]: this comment is out of date now, right?

**Review by danhhz — APPROVED** [review]
I think we should go ahead and get this merged so you don't have to keep it rebased

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (guswynn) [reply src/persist-client/src/examples/persistent_source_example.rs:206]: the important thing to note is there are 2 ways to "block forever":
- Actually blocking the thread, by doing busy-waiting or doing "blocking io"
- "blocking the async fn from making progress", by yielding forever

i think the rust async world leans on only using the word "blocking" for the former to avoid confusion

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (cjubb39) [inline src/persist-client/src/examples/persistent_source_example.rs:489]: Should this be `std::cmp::min`?

Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:489]: I believe so (copy-and-paste error 🙈)

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:153]: is drop just as good for this? that wouldn't break if we switch downgrade_since to be shard-global

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:268]: nit: is this all the bindings that `Timestamper` knows _are durable_?

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:363]: does this still need to return ExternalError?

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:368]: can this internally retry ExternalErrors so it doesn't have to return them?

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:1254]: nit: s/...not good!/closes the shard permanently for writes./

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (danhhz) [inline src/persist-client/src/examples/persistent_source_example.rs:1304]: indeed!

File: src/persist-client/src/impl/machine.rs
Comment (danhhz) [inline src/persist-client/src/impl/machine.rs:172]: hmmm, i'm not sure I agree with this. it's pretty interesting when things retry and it shouldn't be spammy unless we're using unreliable. i'd prefer to keep these at info and re-evaluate when we have data from production logs (or something closer to them)

File: src/persist-client/src/lib.rs
Comment (danhhz) [inline src/persist-client/src/lib.rs:130]: I have a mild preference for the way this is done in the maelstrom PR over this, but I'm willing to change my mind. thoughts? 

https://github.com/MaterializeInc/materialize/pull/12108/commits/aa4bfcdc2dd4580b6e315761ad36a6c47da61d4c

**Review by danhhz — COMMENTED** [review]
I didn't look super closely at the logic for bugs, but the structure looks good and so does the persist usage

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:90]: > edit: oh! is this so you can shut it down and restart it? in that case, maybe we should just abandon the #[test] idea and make this a bin target. that way you can make these flags. i'm happy to fold it in as a subcommand of the persistcli binary being introduced in the maelstrom PR

bingo!

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:153]: Yes, `drop` is actually better.

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:268]: yes

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:363]: It does not, now

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:368]: yep!

File: src/persist-client/src/examples/persistent_source_example.rs
Comment (aljoscha) [reply src/persist-client/src/examples/persistent_source_example.rs:1254]: 👍

File: src/persist-client/src/lib.rs
Comment (aljoscha) [reply src/persist-client/src/lib.rs:130]: I have no preference. I can remove the unreliable code from the example and maybe add it back once your PR is in.

**Pull request merged**
